lib/install/deps.js

```diff
@@ -613,8 +635,12 @@ var earliestInstallable = exports.earliestInstallable = function (requiredBy, tr
   // if this tree location requested the same module then we KNOW it
```

Comment:
> So having written this out it seems clear to me that it should be doing a version match like findRequirement does to see if the version we have in our hands is compatible with the spec in dependencies (and devDependencies) and if it isn't returning null.

Yes, that behavior seems like it helped the original bug to slip through:

My prod deps would throw `lodash@0.9.2` into my root `node_modules` folder.  During `loadDevDeps`, a copy of `lodash@2.0.0` would silently take its place, because `earliestInstallable` wasn't able to realize that the versions were incompatible.

As you mentioned, `earliestInstallable` makes a strong assumption that we _really_ need to install this module _somewhere_, and by falling back to the root `node_modules` folder, guarantees that we do so at all costs.  If `earliestInstallable` crawls all the way up to the root, and finds a conflicting version of the same package already there, it assumes that we _really need_ the package, and overwrites the existing one.  

Unfortunately, `earliestInstallable` has no way of knowing that the versions conflicted, so it had no way of raising a warning.  I don't think we would have a good way of recovering from this situation, but silent failure seems bad if we can avoid it.